File: 2024/Day02/main.py
with open('input.txt', 'r') as f:
    data = f.read()

test = """7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9"""

def diff(data):
    d = []
    for i in range(1, len(data)):
        d.append(data[i]-data[i-1])

    return d


from copy import deepcopy


def get_part_one(data):
    lines = data.split('\n')
    data = [list(map(int, x.split(' '))) for x in lines]

    res = 0
    for d in data:
        d1 =diff(d)
        res += (all( -3 <= x <= -1 for x in d1) or all( 1 <= x <= 3 for x in d1))

    return res

assert get_part_one(test) == 2
print(f'Part 1: {get_part_one(data)}')


def check2(data, depth = 0):
    c_depth = depth+1
    if c_depth > 2:
        return False
    der = diff(data)
    if (all( -3 <= x <= -1 for x in der) or all( 1 <= x <= 3 for x in der)):
        return True

    for i, d in enumerate(data):
        dx = deepcopy(data)
        dx.pop(i)
        res = check2(dx, c_depth)
        if res:
            return True
        
    return False



# part 2 
def get_part_two(data):
    lines = data.split('\n')
    data = [list(map(int, x.split(' '))) for x in lines]

    res = 0
    for d in data:
        d1 =diff(d)
        res += check2(d)

    return res

assert get_part_two(test) == 4
print(f'Part 2: {get_part_two(data)}')

File: 2024/Day02/Rust/src/main.rs
#![allow(non_snake_case)] // Disable non_snake_case warnings for the entire file

use std::fs;
use std::time::Instant;

const TEST: &str = "7 6 4 2 1
1 2 7 8 9
9 7 6 2 1
1 3 2 4 5
8 6 4 4 1
1 3 6 7 9";

fn main() {
    assert_eq!(part_one(TEST), 2);
    let input = fs::read_to_string("../input.txt").expect("Something went wrong reading the file");
    println!("Part 1: {}", part_one(&input));
    assert_eq!(part_two(TEST), 4);
    println!("Part 1: {}", part_two(&input));
}

fn diff(data: &Vec<i32>) -> Vec<i32> {
    let d: Vec<i32> = data.windows(2).map(|w| w[1] - w[0]).collect();
    return d;
}

fn part_one(data: &str) -> u32 {
    let start = Instant::now();
    let mut res = 0;

    let lines = data.lines();
    for line in lines {
        let inter: Vec<&str> = line.split(' ').collect(); // get the vec of str
        let parsed: Vec<i32> = inter.iter().filter_map(|x| x.parse().ok()).collect(); //get the Vec of i32
        let d = diff(&parsed);
        let in_range = d.iter().all(|&x| -3 <= x && x <= -1) || d.iter().all(|&x| 1 <= x && x <= 3);
        if in_range {
            res += 1;
        }
    }

    let duration = start.elapsed();
    println!("Time taken: {:?}", duration);

    return res.try_into().unwrap();
}

fn part_two(data: &str) -> u32 {
    let start = Instant::now();

    let mut res = 0;

    let lines = data.lines();
    for line in lines {
        let inter: Vec<&str> = line.split(' ').collect(); // get the vec of str
        let parsed: Vec<i32> = inter.iter().filter_map(|x| x.parse().ok()).collect(); //get the Vec of i32
        for i in 0..parsed.len() {
            let mut new = Vec::new();
            new.extend_from_slice(&parsed[..i]); // Add elements before index i
            new.extend_from_slice(&parsed[i + 1..]); // Add elements after index i
            let d = diff(&new);
            let in_range =
                d.iter().all(|&x| -3 <= x && x <= -1) || d.iter().all(|&x| 1 <= x && x <= 3);
            if in_range {
                res += 1;
                break;
            }
        }
    }

    let duration = start.elapsed();
    println!("Time taken: {:?}", duration);

    return res.try_into().unwrap();
}
